src/aal/aal_sparc.h

```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+#if defined(__arch64__) // More reliable than __sparc64__
+#  define SNMALLOC_VA_BITS_64
+#else
+#  define SNMALLOC_VA_BITS_32
+#endif
+
+namespace snmalloc
+{
+  /**
+   * Sparc architecture abstraction layer.
+   */
+  class AAL_Sparc
+  {
+  public:
+    /**
+     * Bitmap of AalFeature flags
+     */
+    static constexpr uint64_t aal_features = IntegerPointers;
+
+    static constexpr enum AalName aal_name = Sparc;
+
+#ifdef SNMALLOC_VA_BITS_64
+    /**
+     * Even Ultra-Sparc I supports 8192 and onwards
+     */
+    static constexpr size_t smallest_page_size = 0x2000;
+#else
+    static constexpr size_t smallest_page_size = 0x1000;
+#endif
+
+    /**
+     * On Sparc ideally pause instructions ought to be
+     * optimised per Sparc processor but here a version
+     * as least common denominator to avoid numerous ifdef,
+     * reading Conditions Code Register here
+     */
+    static inline void pause()
+    {
+      __asm__ volatile("rd %%ccr, %%g0" ::: "memory");
+    }
+
+    static inline void prefetch(void* ptr)
+    {
+#ifdef SNMALLOC_VA_BITS_64
+      __asm__ volatile("prefetch [%0], 0" ::"r"(ptr));
+#else
+      UNUSED(ptr);
+#endif
+    }
+
+    static inline uint64_t tick()
+    {
+      uint64_t tick;
+      __asm__ volatile("rd %%asr4, %0" : "=r"(tick));
```

Comment:
SPARC V8 did not define any standard ASRs, only Y/PSR/WIM/TBR, and had this to say about the existing implementations (L.3. Characteristics of Existing Implementations):

> None of the five implementations implements any user or privileged ancillary state registers.

It also said this in the description of RDASR:

> An rs1 value of 1...14 in an RDASR instruction produces undefined results, but does not cause an illegal_instruction trap.

Maybe later V8 implementations added ASR4, but it wasn't until V9 that ASR4 was standardised as TICK, and on V9 `rd %tick, %rd` is valid assembly.

Thus this needs to be guarded by a V9(/V8+, which is just V9 for 32-bit, as Solaris unfortunately conflated the ISA version with the ABI in use) check, and should then also be able to use `tick` rather than `asr4` in the assembly.